x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/queries/SemanticQueryBuilder.java

```diff
@@ -0,0 +1,263 @@
+            if (inferenceResultsList.isEmpty()) {
+                throw new IllegalArgumentException("No inference results retrieved for field [" + fieldName + "]");
+            } else if (inferenceResultsList.size() > 1) {
+                // TODO: How to handle multiple inference results?
+                throw new IllegalArgumentException(
+                    inferenceResultsList.size() + " inference results retrieved for field [" + fieldName + "]"
+                );
+            }
+
+            InferenceResults inferenceResults = inferenceResultsList.get(0);
+            return semanticTextFieldType.semanticQuery(inferenceResults, boost(), queryName());
+        } else {
+            throw new IllegalArgumentException(
+                "Field [" + fieldName + "] of type [" + fieldType.typeName() + "] does not support " + NAME + " queries"
+            );
+        }
+    }
+
+    private SemanticQueryBuilder doRewriteGetInferenceResults(QueryRewriteContext queryRewriteContext) {
+        if (inferenceResults != null) {
+            return this;
+        }
+
+        if (inferenceResultsSupplier != null) {
+            InferenceServiceResults inferenceResults = inferenceResultsSupplier.get();
+            return inferenceResults != null ? new SemanticQueryBuilder(this, null, inferenceResults) : this;
+        }
+
+        ResolvedIndices resolvedIndices = queryRewriteContext.getResolvedIndices();
+        if (resolvedIndices == null) {
+            throw new IllegalStateException(
+                "Rewriting on the coordinator node requires a query rewrite context with non-null resolved indices"
+            );
+        } else if (resolvedIndices.getRemoteClusterIndices().isEmpty() == false) {
+            throw new IllegalArgumentException(NAME + " query does not support cross-cluster search");
+        }
+
+        String inferenceId = getInferenceIdForForField(resolvedIndices.getConcreteLocalIndicesMetadata().values(), fieldName);
+        SetOnce<InferenceServiceResults> inferenceResultsSupplier;
+        if (inferenceId != null) {
+            InferenceAction.Request inferenceRequest = new InferenceAction.Request(
+                TaskType.ANY,
+                inferenceId,
+                null,
+                List.of(query),
+                Map.of(),
+                InputType.SEARCH,
+                InferModelAction.Request.DEFAULT_TIMEOUT_FOR_API
+            );
+
+            inferenceResultsSupplier = new SetOnce<>();
+            queryRewriteContext.registerAsyncAction(
+                (client, listener) -> executeAsyncWithOrigin(
+                    client,
+                    ML_ORIGIN,
+                    InferenceAction.INSTANCE,
+                    inferenceRequest,
+                    listener.delegateFailureAndWrap((l, inferenceResponse) -> {
+                        inferenceResultsSupplier.set(inferenceResponse.getResults());
```

Comment:
Should we check for `warning` inference results size & supported kinds early and fail?

Reply:
This is a good point and also made me realize we should probably convert from `InferenceServiceResults` to `InferenceResults` in `doRewriteGetInferenceResults`, which is executed on the coordinator node.

As currently written, each shard will do this conversion individually (when `doRewriteBuildSemanticQuery` is executed). By checking & converting earlier, we prevent duplicate work and do (some of) the error checking in the coordinator node, avoiding an unnecessary serialization/deserialization when the inference results are clearly invalid.

@carlosdelest WDYT?

Reply:
Makes sense to me - if we can avoid doing work that can be preempted in the coordinating node, let's do it

Reply:
I addressed this, but it required more changes than I anticipated. This is due to the fact that prior to addressing this, the query enforced that `inferenceResults` is always non-null after re-writing on the coordinator node. With the changes in https://github.com/elastic/elasticsearch/pull/108483/commits/bceeb842e2c884037cd0c056800f31a7d8d8d536, this is no longer the case, which created a couple more edge cases.

I also want to add a test to `SemanticQuerBuilderTests` to validate that the query can be serialized after re-writing on the coordinator node, particularly when no inference ID can be looked up for the field queried. This is not tested by the YAML tests because they all run on single-node clusters.

Reply:
Added the test in https://github.com/elastic/elasticsearch/pull/108483/commits/27dff4b8cbed2e2d0d20ba95ad1b5773af8d660e

Reply:
Thanks for adding the tests, Mike!